IDinsight/ask-a-question
# [128] - Google cloud TTS and STT functionality

Reviewer: @lickem22 
Estimate: 20 mins

---

## Ticket

Fixes: [JIRA_TICKET_LINK](https://idinsight.atlassian.net/browse/AAQ-768?atlOrigin=eyJpIjoiMzRkYjdkMDcwNDRkNGM0NmIxOWYzMTQ4MmViZDA0MTIiLCJwIjoiaiJ9)

## Description
This pull request  implements the Google Cloud `text-to-speech (TTS`) and `speech-to-text (STT)` external APIs in the existing AAQ  workflow, alongside the existing in-house models.

### Goal
The goal of this PR is to provide users with the option to use either the in-house models or the new Google Cloud `TTS` and `STT` models, depending on their specific use cases and requirements.

### Changes
- Added the ability to select between external APIs or in-house models based on the `CUSTOM_SPEECH_ENDPOINT` environment variable.
- introduced new external voice components in the `core_backend/app/question_answer/speech_components` directory.

### Future Tasks 
- Determine the best approach to detect the language for the external STT model.
- Explore the possibility of an automated way to map languages to their corresponding language codes.
 
## How has this been tested?
docker-compose
swagger UI
Unit Tests (pytests)

### How to test this?
1. Ensure that the `CUSTOM_SPEECH_ENDPOINT` environment variable is not set, to access the external APIs, and that your `GCP credentials `are properly configured.
2. Run the command `docker compose -f docker-compose.yml -f docker-compose.dev.yml -p aaq-stack watch` (the `speech_service` container is not needed for the external APIs).
3. Test the functionality by accessing the endpoint and viewing the generated file through the `generated URL`.

## Checklist

Fill with `x` for completed.
